### ISSUE TITLE
mirror_whitelist inheritance

### ISSUE THREAD

**Issue by fschulze** [devpi/devpi#725]
A problem with ``mirror_whitelist`` inheritance was brought to my attention which showed that my own expectation of how ``mirror_whitelist`` currently works was wrong.

There is a common need to upload binary wheels for packages which don't have the required ones, but at the same time have private packages. My suggestion for a solution was this:

1. create an index "org/wheels" which has ``bases=root/pypi`` and ``mirror_whitelist=*``
2. create an index "org/private" which has ``bases=org/wheels`` and the default empty ``mirror_whitelist``

My expectation was, that any private package uploaded in "org/private" would be protected by its whitelist. Unfortunately that is not the case, as the whitelist is checked for each index in the chain.

My proposed solution is to only use ``mirror_whitelist`` from the first index in the inheritance chain.

The problem with that is, that existing setups might break with that change.

So, two questions:
1. Is my proposed change correct or did I overlook something?
2. Is this a bugfix, or does it require a major version bump?

cc @r2dan @StephanErb

**Comment by r2dan** [comment]
@fschulze we need to gather some feedback from other teams. For me it looks fine - way more explicit. We'll get back to you soon.

**Comment by fschulze** [comment]
A better solution would probably be to narrow down the whitelist by intersecting them in each step.

With my first proposal the "org/private" index will overwrite the ``mirror_whitelist`` of "org/wheels". But when you add a third "org/dev" index with ``bases=org/private`` then setting ``mirror_whitelist=*`` would remove the protection from "org/private". With the intersecting the protection stays in place.

**Comment by beamerblvd** [comment]
Well we also want to have an `org/dev` that inherits from `org/private` with `mirror_whitelist='*'`, so I would prefer for that to work as well. If your concerned about breaking existing behavior for people depending on it, what about a new explicit value, such as `mirror_whitelist='~*'`, that invokes the new behavior?

**Comment by fschulze** [comment]
@beamerblvd you mean you *want* a way to overwrite and not intersect?

Adding such a prefix would affect performance, because we would have to parse the whitelist each time instead of just using it as a set.

Maybe a new setting which selects the inherit behaviour would be better. Existing indexes would use the old behaviour and new indexes would select a safe default. The modes would be ``union`` (current behaviour), ``intersect`` and ``first`` (could maybe use a better name). The option could be named ``mirror_whitelist_inheritance`` and the new default would be ``intersect`` and existing indexes would use ``union``.

**Comment by fschulze** [comment]
``overwrite`` seems better than ``first``

**Comment by beamerblvd** [comment]
A new setting sounds smart. In case I didn’t state it correctly, I need:

- `org/wheels` to extend `root/pypi` and packages that exist in both can intermingle versions
- `org/private` to extend `org/wheels` and any package in `org/private` completely overwrites and hides a package with the same name in either `org/wheels` or `org/private`
- `org/dev` to extend `org/private` and packages that exist in both can intermingle versions

So if you had this:

- `root/pypi`: `django-2.2.tar.gz`
- `root/pypi`: `permissions-0.1.tar.gz`
- `org/wheels`: `django-2.2-py3-none.any.whl`
- `org/private`: `permissions-5.7-py2.py3-none-any.whl`
- `org/dev`: `permissions-5.8.dev.tar.gz`

The behavior would be:

- If you queried `org/wheels` for `django`, you'd see `django-2.2.tar.gz` and `django-2.2-py3-none.any.whl`
- If you queried `org/private` for `django`, you'd see `django-2.2.tar.gz` and `django-2.2-py3-none.any.whl`
- If you queried `org/dev` for `django`, you'd see `django-2.2.tar.gz` and `django-2.2-py3-none.any.whl`
- If you queried `org/wheels` for `permissions`, you'd see `permissions-0.1.tar.gz`
- If you queried `org/private` for `permissions`, you'd see `permissions-5.7-py2.py3-none-any.whl` and you would **_NOT_** see `permissions-0.1.tar.gz` (completely hidden)
- If you queried `org/dev` for `permissions`, you'd see `permissions-5.7-py2.py3-none-any.whl` and `permissions-5.8.dev.tar.gz` and you would **_NOT_** see `permissions-0.1.tar.gz` (completely hidden)

**Comment by beamerblvd** [comment]
Also: How can I help?

**Comment by beamerblvd** [comment]
I'm uncomfortable with the name "intersect," because when dealing with set-wise math, it means something different, but you're also using "union" to mean the same as set-wise math, which is confusing.

In set-wise math:
- The union of two sets is a set that contains everything that is in EITHER set
- The intersection of two sets is a set that contains everything that is in BOTH sets

I can't imagine a use-case where you would want to intersect an index and its base. That defeats the purpose of index inheritance entirely.

How about this?
- `union` (current behavior, what the `org/wheels` I described above does, default until maybe a major version)
- `overwrite` (inherit everything that doesn't conflict, overwrite everything that does conflict)
- `union-previous` (union with the previous index(es) and what it inherits, but if the previous index(es) is(are) `overwrite`, respect that and don't union with conflicting packages in the base of the previous index(es))

**Comment by beamerblvd** [comment]
Sorry for the flood of three replies above ^^ 😄

**Comment by fschulze** [comment]
@beamerblvd Your usecase would be the new default of ``intersect``.

In your case "org/wheels" would have ``mirror_whitelist=*``, "org/private" and "org/dev" would both have ``mirror_whitelist=`` and ``mirror_whitelist_inheritance=intersect``. Due to the intersection of the whitelist at each step, the effective whitelist would always be empty and thus not whitelisting anything.

The current behaviour is the same as when we do a union of the whitelist at each step, because the union of the empty sets with the ``{'*'}`` from "org/wheels" results in the effective ``{'*'}``.

With ``mirror_whitelist_inheritance=overwrite`` and ``mirror_whitelist=*`` on "org/dev" you *would* get all three ``permissions`` packages.

So in essence the set operations are on the ``mirror_whitelist``, *not* the index contents. The ``mirror_whitelist_inheritance`` setting is only read from and only affects the current index.

Now that I've written this down I don't see a difference in ``union`` and ``overwrite``, unless I'm missing something.

So to summarize:
1. we would have two modes: ``union`` and ``intersection``
2. the default for new indexes is ``intersection`` and is explicitly set on creation
3. when the setting is missing the default is ``union`` to keep existing installations working as they are

**Comment by beamerblvd** [comment]
Okay, that makes sense, if the terms `union` and `intersection` are referring to the whitelist settings themselves instead of referring to the  index contents.

**Comment by fschulze** [comment]
@beamerblvd your examples are helpful to write tests and the discussion also helped clearing things up. I'll let you know when there is something to test. It won't get into 5.2.0 though.

**Comment by beamerblvd** [comment]
Bummer, but understood. I'll help in any way I can, just let me know. 

What will be the consequences of deploying with the current settings and then changing `mirror_whitelist_inheritance` once it's released? Will it be possible? Or will it break things? Will the existence of replicas replicating from the master affect whether this setting can be changed after upgrade?

**Comment by fschulze** [comment]
@beamerblvd There are no consequences for existing data. As soon as you change the setting after its available the new feature will work. The replicas will just update once the new setting is applied. In general replicas should be updated before the master though, so code changes are already available when the master uses them.

**Comment by beamerblvd** [comment]
Makes sense. That's how we've handled MySQL upgrades in the past: Upgrade all the replicas, then upgrade the master.

**Comment by beamerblvd** [comment]
Have you had a chance to look into this any recently? Any chance we could get this in 5.4? :-)

**Comment by fschulze** [comment]
@beamerblvd no, I haven't had time to look into this. It's a fair amount of work as well to add the tests for various cases. If you would want to try to make a PR, I can give you some pointers. If you are able to get a budget for implementing this, then you can contact us via https://merlinux.eu for details.

**Comment by fschulze** [comment]
Fix will be in devpi-server 6.0.0.
